internal/gh/gh.go

```diff
@@ -9,13 +10,28 @@ import (
 
 // CreatePullRequest creates a pull request.
 func CreatePullRequest(args ...string) error {
+	err := ensureGHInstalled()
+	if err != nil {
+		return err
+	}
+
 	cmd := exec.Command("gh", args...)
 	cmd.Stdin = os.Stdin
 	cmd.Stdout = os.Stdout
 	cmd.Stderr = os.Stderr
-	err := cmd.Run()
+	err = cmd.Run()
 	if err != nil {
 		return fmt.Errorf("running gh command with args %q: %w", strings.Join(args, " "), err)
 	}
 	return nil
 }
+
+func ensureGHInstalled() error {
+	cmd := exec.Command("which", "gh")
+	err := cmd.Run()
+	if err != nil {
+		fmt.Println("🤓 This command requires the gh cli.\nSee: https://github.com/cli/cli")
+		return errors.New("missing gh cli dependency")
+	}
+	return nil
```

Comment:
I was thinking about something like this untested code:

```go
func ensureGHInstalled() error {
	return ensureToolIsInstalled("gh", "This command requires the gh cli.\\nSee: https://github.com/cli/cli")
}

func ensureToolIsInstalled(toolName string, extraHelp string) error {
	cmd := exec.Command("which", toolName)
	err := cmd.Run()
	if err == nil {
		return nil
	}
	if runtime.GOOS == `darwin` {
		fmt.Printf("Installing '%s' with brew\n", toolName)
		cmd = exec.Command("brew", "install", toolName)
		err = cmd.Run()
	}
	if err != nil {
		fmt.Printf("Ensure '%s' is in your PATH. %s\n", toolName, extraHelp)
		return errors.New(fmt.Sprintf("missing %s dependency", toolName))
	}
	return nil
}

```

Reply:
Very interesting!  However, not sure about automatic "lazy" installation when the command is run...  Would maybe more implement that in `joy setup`, especially as some dependencies like `jac` require further setup (clone people repo).  Maybe we could suggest user to run `joy setup` instead, or prompt whether to perform that installation (and then execute just that portion of the setup process)? I will create a separate ticket for that, instead of adding all that logic within this PR (following advice from a great collaborator who just joined our team!)

Reply:
@j-martin please review that ticket and let me know what you think:
https://nestoca.atlassian.net/browse/DEVOPS-1938

Reply:
Here's the ticket! 🙂
https://nestoca.atlassian.net/browse/DEVOPS-1938

Reply:
The ticket looks good. I could imagine installing the tools dynamically could be done in this PR, but it is also reasonable not to pack too much in a PR.

That got me thinking about `gh` precisely, now `joy` would also depend on `gh` being configured properly.

We could check the auth status:

```
$ gh auth status
github.com
  ✓ Logged in to github.com as j-martin (keyring)
  ✓ Git operations for github.com configured to use ssh protocol.
  ✓ Token: gho_************************************
  ✓ Token scopes: admin:public_key, gist, read:org, repo
```

Reply:
That's what I ended up doing, let me know if that's what you had in mind:
https://github.com/nestoca/joy/commit/ae2dd774b09d058b889811da90945af345888676

Reply:
We really need to check the exit code.

```
$ gh auth status
github.com
  ✓ Logged in to github.com as j-martin (keyring)
  ✓ Git operations for github.com configured to use ssh protocol.
  ✓ Token: gho_************************************
  ✓ Token scopes: admin:public_key, gist, read:org, repo
$ echo $?
0

# Deleted the token from my local keychain
$ gh auth status
github.com
  X github.com: authentication failed
  - The github.com token in /Users/jm/.config/gh/hosts.yml is no longer valid.
  - To re-authenticate, run: gh auth login -h github.com
  - To forget about this host, run: gh auth logout -h github.com
$ echo $?
1
```

Reply:
The code actually works, because it checks for the presence of `Token: XXXXXX`.  I ignored exit code / error, to focus on that check instead, because the two checks were redundant, but I could rely solely on `gh auth status` returning 0 exit code (plus the scopes check).